### ISSUE TITLE
New iOS web content restrictions block cordova-plugin-meteor-webapp's access to localhost, app won't run

### ISSUE THREAD

**Issue by graemian** [meteor/meteor#13339]
Recent versions of iOS block access to localhost when web content is restricted by Screen Time. Since Meteor uses http://localhost to serve Cordova content, users see this error page when opening the app:

![Finder 2024-09-09 18 50 30](https://github.com/user-attachments/assets/cd3e0661-cd6a-4da8-8453-6ed49c286fed)

Tapping "Allow Website" does nothing.

This happens when web content is restricted to approved sites only:

![Finder 2024-09-09 18 21 46](https://github.com/user-attachments/assets/027902f0-4275-453a-85e5-ff9e00f4c588)

If I add "localhost" as an approved site, it works:

![Preview 2024-09-09 18 25 18](https://github.com/user-attachments/assets/f328c2f0-25fd-4d01-9933-fbc08da9580e)

(You can type anything into the Title field, Bomad is just the name of my app)

This happens on iPadOS 17.5.1 and iOS 17.6.1, but also on iOS 16 (see https://forums.developer.apple.com/forums/thread/732008).

I've tested on Meteor 2.16, but I'm quite sure the problem exists for Meteor 3 too.

This seems to be a possible solution:

> For those who are still stumbling upon this, newer versions of cordova-plugin-ionic-webview now support using the app scheme instead of serving over http://localhost/. Using the app scheme will allow the app to not get blocked by iOS’ Content Restriction. Hope this helps!

(source: https://stackoverflow.com/a/60998000/1420157)

**How to reproduce:**

1. Enable web content restrictions: 

Go to Settings > Screen Time > Content & Privacy Restrictions > Store, Web, Siri & Game Center Content ("Content Restrictions" on iOS 16) > Web Content

and choose "Only Approved Websites"

2. Install any Meteor Cordova app (you can use mine: https://apps.apple.com/us/app/bomad-kids-money-tracker/id6446114229) 

3. Open the app, you will see the error page above

**Comment by graemian** [comment]
I just realized that pressing "Allow Website" on the error page:

![Finder 2024-09-10 11 17 12](https://github.com/user-attachments/assets/009e1047-c1a4-4074-baa4-5e2aeef9da8f)

adds "localhost" to the list of approved sites. 

However, it doesn't trigger a reload, so the error remains on the screen. It also doesn't give any other kind of feedback, so it seems like nothing happened.

If you close the app and re-open it, it works. However, it's not obvious that you need to do this.

**Comment by nachocodoner** [comment]
So, were you able to test the fix using this plugin, cordova-plugin-ionic-webview? I am not sure which options we have on this except to recommend the usage of this plugin, or integrate it automatically and have a different value from localhost for the hostname.

**Comment by graemian** [comment]
No, I haven’t tested this plugin. Assuming it works, I don’t think you can simply recommend using it instead. As I understand, cordova-plugin-meteor-webapp does all kinds of other things too, like handling app updates (see https://www.npmjs.com/package/cordova-plugin-meteor-webapp).

I suspect the easiest fix would be to see how cordova-plugin-ionic-webview avoids using localhost, and then to try to update cordova-plugin-meteor-webapp to do the same. I don’t think that will be trivial. I'm a bit intimated by all the C++ I see inside of cordova-plugin-meteor-webapp.

**Comment by nachocodoner** [comment]
Taking a note on this issue. I feel this one is kind of urgent, right? Any iOS app will start experiencing this issue.

Even having this workaround, https://github.com/meteor/meteor/issues/13339#issuecomment-2340129665, it seems so annoying.

Also, have you been able to reproduce locally in local physical device development?

**Comment by graemian** [comment]
Yeah, any iOS app with web content restrictions enabled will experience this problem on iOS 17.5 and above, so I would say it's quite urgent.

Worse still, I think that a blank screen is sometimes shown instead of that error screen ("You cannot browse this page at “localhost” because it is restricted"). I'm pretty sure I've experienced this. More details here:

>  Also, removing and adding the ios platform helped a bit. Instead of showing a blank white screen after splash it showed a message saying that the site is restricted and offers the user to allow it.

https://stackoverflow.com/a/50197914/1420157


I have just tested and I can replicate this problem locally on my physical iPad by running my app from Xcode.

Tx!

**Comment by nachocodoner** [comment]
I'm running a basic Meteor app on an iOS device with version 17.5. I'm not seeing either the white crash page or the restricted message.

Could you assist me in reproducing the issue?

Can you replicate it with a simple React app created using `meteor create my-app --react` and using iOS emulator?

If you have a specific app setup, could you share it in a repository so I can try it myself?

Note: I just realized you reproduced on a physical device. I am getting my ready, I will get back on the result of reproduction.

Worked for me in my physical device on 17.5.1 version, [demo](https://github.com/user-attachments/assets/46c3035e-00fb-4571-9a2d-9243191fc674)

**Comment by graemian** [comment]
Let me test with a new project, will let you know shortly.

Are you sure that you enabled web content restrictions as below?

> Go to Settings > Screen Time > Content & Privacy Restrictions > Store, Web, Siri & Game Center Content ("Content Restrictions" on iOS 16) > Web Content and choose "Only Approved Websites"

**Comment by graemian** [comment]
I just tested it again with a new app ("meteor create my-app --react") on my iPad with iOS 17.6.1 and I got the "Restricted Site" message.

I got the same error last week with my app when the iPad was still on iOS 17.5.1.

Here's roughly what I did:

1. meteor create my-app --react
2. meteor add-platform ios
3. On the iPad, go to Settings > Screen Time > Content & Privacy Restrictions > Store, Web, Siri & Game Center Content ("Content Restrictions" on iOS 16) > Web Content and choose "Only Approved Websites"
4. "meteor run ios-device" then run on device from Xcode

I'll test on an iPhone and report back.

**Comment by graemian** [comment]
I just tested a new app on an iPhone with iOS 16.2 and it shows this error when content restrictions are enabled:

![image](https://github.com/user-attachments/assets/e24e3f12-d8dc-4251-86ed-276c9f71926d)

**Comment by nachocodoner** [comment]
Reproduced, thank you for your support.

I missed that step because I misunderstood the issue. I thought it was a default config for the new iOS version, but it seems you need to enable the "Only Approved Websites" setting to trigger the problem. However, after checking the Apple forums link you shared, I see some devices have these content filters set by default. So, this issue affects some cases, but not all.

The best workarounds are as mentioned above: removing the security setting if possible, or adding localhost to the allowed websites.

Good that we could reproduce this locally. We can add this issue to our backlog now. Regarding the solution, we can look into how `cordova-plugin-ionic-webview` avoids using localhost and apply that in our webview package as suggested above. Another option is to add and configure `cordova-plugin-ionic-webview` as a default for Meteor apps, assuming this approach works, since we haven't confirmed that yet. But if we add this plugin we will cause sort of incompatibilities on existing Meteor cordova apps, as we already handle ios webview I suppose.

**Comment by graemian** [comment]
Thanks. It’s quite a pain in my case since many kids use my app and parents tend to restrict websites on their kids devices.

I wondered if switching to 127.0.0.1 might fix the problem, but it doesn’t seem to be the case - see https://github.com/readium/swift-toolkit/issues/325#issuecomment-1652158519

Let me know if there’s anything more I can do to help.

**Comment by nachocodoner** [comment]
We have been addressing fixes on Cordova, concentrating on Meteor 3. I'm unclear about the urgency of this issue, as we have only received one report. It would be helpful to find a workaround before implementing a proper and "complex" fix.

Our main focus is on Meteor 3 and other pressing priorities, such as performance. I’d like to better understand the overall impact of this issue. It appears that only users who intentionally activate this feature are impacted, like parents with their children's phones. Have you encountered a mobile device or iOS version that comes with website restrictions enabled by default?

**Comment by graemian** [comment]
Setting up content restrictions is part of the Apple child account creation process, so I think quite a few kids have it enabled:

<img width="271" alt="image" src="https://github.com/user-attachments/assets/daf94107-01ad-40fc-865b-7046ec23dcf0">

(from "[How to create a child account with Family Sharing on iPhone or iPad | Apple Support](https://youtu.be/8siFNH-brM4?t=63)")

So this is quite a big problem for any child or family app.

Also, I suspect that content restrictions are enabled for iPads that are used for specific purposes in companies. For example, an iPad used for inventory management will have content restrictions enabled and so wouldn't be able to run an inventory management app built with Meteor.

I'll let you know if I think of anything else.

I wish I could find a workaround. It's especially frustrating that nothing happens when you tap "Allow website". And because the problem occurs right at startup, you can't communicate anything to the user.

I'm busy updating my iPad to iOS 18 - maybe it's better somehow. I'll let you know.

**Comment by graemian** [comment]
Same problem on iOS 18 🙁

**Comment by graemian** [comment]
Still relevant, still struggling with app reviews about this:

<img width="687" alt="Image" src="https://github.com/user-attachments/assets/5abcf6ae-5d4f-4347-ba03-497ffe3ae341" />
